src/http/client.c

```diff
@@ -198,24 +317,94 @@ static void recv_handler(struct mbuf *mb, void *arg)
 
  out:
 	req_close(req, err, msg);
-	mem_deref(req);
 	mem_deref(msg);
+
+	if (err)
+		mem_deref(conn);
 }
 
 
 static void close_handler(int err, void *arg)
 {
-	struct http_req *req = arg;
+	struct conn *conn = arg;
 
-	if (req->srvc > 0 && !req->data) {
+	try_next(conn, err ? err : ECONNRESET);
+}
 
-		err = req_connect(req);
-		if (!err)
-			return;
+
+static bool conn_cmp(struct le *le, void *arg)
+{
+	const struct conn *conn = le->data;
+	const struct http_req *req = arg;
+
+	if (!sa_cmp(&req->srvv[req->srvc], &conn->addr, SA_ALL))
+		return false;
+
+	if (req->secure != !!conn->sc)
+		return false;
+
+	return conn->req == NULL;
```

Comment:
I guess conn->req is set if it has a pending request, and conn->req==NUL if the connection
is free to be reused by other requests?

perhaps we should add a small comment about this.